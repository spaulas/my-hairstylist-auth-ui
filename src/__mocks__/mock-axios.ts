export const mockPost = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockGet = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockPut = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockDelete = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockPatch = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

const mockAxios = jest.fn().mockReturnValue({
  post: mockPost,
  get: mockGet,
  put: mockPut,
  patch: mockPatch,
  delete: mockDelete,
});

jest.mock("../config/HttpClient/index.ts", () => ({
  __esModule: true,
  default: mockAxios,
}));
