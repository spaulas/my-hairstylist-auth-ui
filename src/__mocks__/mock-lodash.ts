export const mockDebounce = jest.fn((fn) => console.log('MOCK DEBOUTNCE'));

jest.mock("lodash.debounce", () => ({
  ...jest.requireActual("lodash.debounce"),
  debounce: mockDebounce,
}));
