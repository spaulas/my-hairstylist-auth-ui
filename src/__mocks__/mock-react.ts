export const mockUseState = jest.fn();
export const mockUseCallBack = jest.fn((fn) => fn);

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: mockUseState,
  useCallback: mockUseCallBack,
}));
