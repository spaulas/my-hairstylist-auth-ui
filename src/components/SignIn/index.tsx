import { useState } from "react";
import Form from "@components/generics/Form";
import EmailField from "@components/generics/Form/Field/Email";
import PasswordField from "@components/generics/Form/Field/Password";
import useAuth from "@hooks/useAuth/index.hooks";

const SignIn = () => {
  const { signIn } = useAuth();
  const [fields, setFields] = useState<Record<string, string>>({
    password: "",
    email: "",
  });
  
  const handleSignIn = () => {
    signIn(fields);
  };

  return (
    <Form
      handleSubmit={handleSignIn}
      buttonTitle="Login"
      showForgotPassword
      link={{
        message: "Don't have an account?",
        title: "Sign up",
        url: "/auth/sign-up",
      }}
    >
      {(isFocused: Record<string, boolean>, setIsFocused: any) => (
        <>
          <EmailField
            email={fields.email}
            setEmail={(value: string) => setFields({ ...fields, email: value })}
            isFocused={isFocused.email}
            setIsFocused={(value: any) =>
              setIsFocused({ ...isFocused, email: value })
            }
          />
          <PasswordField
            password={fields.password}
            setPassword={(value: string) =>
              setFields({ ...fields, password: value })
            }
            isFocused={isFocused.password}
            setIsFocused={(value: any) =>
              setIsFocused({ ...isFocused, password: value })
            }
          />
        </>
      )}
    </Form>
  );
};

export default SignIn;
