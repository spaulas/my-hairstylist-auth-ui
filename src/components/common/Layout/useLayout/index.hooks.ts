import { useSelector, useDispatch } from "react-redux";
import { RootReducerState } from "@type/global";
import layoutActions from "@actions/layout/layout.action";

const useLayout = () => {
  const dispatch = useDispatch();

  const { isBackgroundFocused } = useSelector(
    ({ Layout }: RootReducerState) => {
      return { isBackgroundFocused: Layout.isBackgroundFocused };
    }
  );

  const setBackgroundFocus = (focused: boolean) => {
    dispatch(layoutActions.setBackgroundFocus(focused));
  };

  return { isBackgroundFocused, setBackgroundFocus };
};

export default useLayout;
