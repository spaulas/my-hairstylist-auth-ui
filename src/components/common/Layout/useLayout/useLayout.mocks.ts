export const mockUseLayout = jest.fn();

jest.mock("./index.hooks.ts", () => mockUseLayout);
