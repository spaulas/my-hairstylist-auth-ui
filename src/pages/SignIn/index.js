import React from "react";

const SignIn = () => {
  return <div>SIGN IN PAGE</div>;
};

export default SignIn;
