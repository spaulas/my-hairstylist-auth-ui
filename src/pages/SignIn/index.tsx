import React from "react";
import Layout from "@components/common/Layout";
import SignInForm from "@components/SignIn";

const SignIn = () => {
  return (
    <Layout>
      <SignInForm />
    </Layout>
  );
};

export default SignIn;
