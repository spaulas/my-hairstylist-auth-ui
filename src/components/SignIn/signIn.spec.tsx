import "@testing-library/jest-dom/extend-expect";
import { mockUseState } from "../../__mocks__/mock-react";
import { mockUseAuth } from "@hooks/useAuth/useAuth.mocks";
import { mockUseLayout } from "@components/common/Layout/useLayout/useLayout.mocks";
import { fireEvent, render, screen } from "@testing-library/react";
import SignIn from ".";

describe("SignIn", () => {
  const mockSignIn = jest.fn();

  const fields = {
    password: "",
    email: "",
  };
  const mockSetFields = jest.fn();

  beforeEach(() => {
    mockUseAuth.mockReturnValue({ signIn: mockSignIn });
    mockUseState.mockReturnValue([fields, mockSetFields]);
    mockUseLayout.mockReturnValue({
      isBackgroundFocused: false,
      setBackgroundFocus: jest.fn(),
    });
  });

  it("renders correctly", () => {
    const { container } = render(<SignIn />);
    expect(container).toMatchSnapshot();
  });

  describe("renders correct fields", () => {
    beforeEach(() => {
      render(<SignIn />);
    });

    it("button", () => {
      const button = screen.getByText("Login");
      expect(button).toBeVisible();
    });

    it("forgotPassword", () => {
      const forgotPassword = screen.getByText("Forgot your password?");
      expect(forgotPassword).toBeVisible();
    });

    it("link message", () => {
      const linkMessage = screen.getByText("Don't have an account?");
      expect(linkMessage).toBeVisible();
    });

    it("link title", () => {
      const linkTitle = screen.getByText("Sign up");
      expect(linkTitle).toBeVisible();
    });

    it("email field", () => {
      const emailTitle = screen.getByText("Email");
      expect(emailTitle).toBeVisible();
    });

    it("password field", () => {
      const passwordTitle = screen.getByText("Password");
      expect(passwordTitle).toBeVisible();
    });
  });

  it("link should have correct url", () => {
    render(<SignIn />);
    const linkTitle = screen.getByText("Sign up");
    expect(linkTitle).toHaveAttribute("href", "/auth/sign-up");
  });

  describe("should call setFields when", () => {
    it("email field changes", () => {
      render(<SignIn />);

      const emailInput = screen.getByTestId("email");
      const email = "email@email.com";
      fireEvent.change(emailInput, { target: { value: email } });

      expect(mockSetFields).toHaveBeenCalledWith({ ...fields, email });
    });

    it("password field changes", () => {
      render(<SignIn />);

      const passwordInput = screen.getByTestId("password");
      const password = "email@email.com";
      fireEvent.change(passwordInput, { target: { value: password } });

      expect(mockSetFields).toHaveBeenCalledWith({ ...fields, password });
    });
  });

  it("calls signIn when button is clicked", () => {
    render(<SignIn />);

    const button = screen.getByText("Login");
    fireEvent.click(button);

    expect(mockSignIn).toHaveBeenCalledTimes(1);
  });
});
