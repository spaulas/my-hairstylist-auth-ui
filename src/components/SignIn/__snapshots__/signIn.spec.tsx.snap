// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn renders correctly 1`] = `
<div>
  <div
    class="form"
  >
    <h1
      class="form--title "
    >
      My Hairstylist
    </h1>
    <div
      class="form--item"
    >
      <p
        class="form--label "
      >
        Email
      </p>
      <input
        class="form--input "
        data-testid="email"
        id="email"
        name="email"
        type="email"
        value=""
      />
      <p
        class="error error__visible"
      >
        Please add a valid email
      </p>
    </div>
    <div
      class="form--item"
    >
      <p
        class="form--label "
      >
        Password
      </p>
      <input
        class="form--input "
        data-testid="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <p
        class="error error__visible"
      />
    </div>
    <button
      class="primary"
    >
      Login
    </button>
    <a
      class="link form--links-forgot-password"
      href=""
    >
      Forgot your password?
    </a>
    <p
      class="form--links-sign-up"
    >
      Don't have an account?
       Â 
      <a
        class="link "
        href="/auth/sign-up"
      >
        Sign up
      </a>
    </p>
  </div>
</div>
`;
